libtenzir/builtins/functions/string.cpp

```diff
@@ -211,6 +211,136 @@ class trim : public virtual function_plugin {
+              if (subject_array.IsNull(i) || length_array.IsNull(i)) {
+                check(b.AppendNull());
+                continue;
+              }
+              auto str = subject_array.GetView(i);
+              auto target_length = length_array.Value(i);
+              if (target_length <= 0) {
+                check(b.Append(""));
+                continue;
+              }
+              // For simple string length, we can use the string view's size for
+              // ASCII or count UTF-8 characters manually.
+              auto str_length = int64_t{0};
+              auto ptr = str.data();
+              auto end = ptr + str.size();
+              while (ptr < end) {
+                // Skip UTF-8 continuation bytes (10xxxxxx)
+                if ((*ptr & 0xC0) != 0x80) {
+                  str_length++;
+                }
+                ptr++;
+              }
+              if (str_length >= target_length) {
+                // String is already long enough.
+                check(b.Append(str));
+                continue;
+              }
+              // Validate pad character is single character.
+              int64_t pad_char_length = 0;
+              ptr = pad_char->data();
+              end = ptr + pad_char->size();
+              while (ptr < end) {
+                if ((*ptr & 0xC0) != 0x80) {
+                  pad_char_length++;
+                }
+                ptr++;
+              }
+              if (pad_char_length != 1) {
+                diagnostic::warning("`{}` expected single character for "
+                                    "padding, "
+                                    "but got `{}` with length {}",
+                                    name, *pad_char, pad_char_length)
+                  .primary(subject_expr)
+                  .emit(ctx);
+                check(b.AppendNull());
+                continue;
+              }
+              // Calculate padding needed.
+              auto padding_needed
+                = static_cast<size_t>(target_length - str_length);
+              std::string result;
+              result.reserve(str.size() + padding_needed * pad_char->size());
+              if (pad_left) {
+                // Pad on the left
+                for (size_t j = 0; j < padding_needed; ++j) {
+                  result += *pad_char;
+                }
+                result += str;
+              } else {
+                // Pad on the right
+                result = str;
+                for (size_t j = 0; j < padding_needed; ++j) {
+                  result += *pad_char;
+                }
+              }
+              check(b.Append(result));
+            }
+          },
+          [&](const auto&, const auto&) {
+            // Type mismatch
+            diagnostic::warning("`{}` expected (string, int), but got ({}, {})",
+                                name, subject.type.kind(), length.type.kind())
+              .primary(subject_expr)
+              .emit(ctx);
```

Comment:
We normally report type errors for a single parameter instead of combining everything. Then you can also use the correct `primary` (it's not correct right now if `length` has the bad type).